### PR TITLE
[trainer, perf] fix: apply HSDP all-reduce control to custom trainers

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

PR #781 added `_configure_hsdp_allreduce()` to avoid redundant synchronization over the HSDP replicated dimension during gradient accumulation. `BaseTrainer.train_step()` uses the helper, but the custom train loops in the text, VLM, DPO, and DiT trainers bypassed it.

This PR applies the existing helper in those train loops so replica all-reduce is disabled for the first accumulation micro-step and restored for the final micro-step. The DiT offline-embedding path remains unchanged because it does not execute model forward/backward.

Related: #781

### Checklist Before Starting

- Search for relative PRs/issues and link here: #781
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

Current head after rebasing onto `main`:

```text
make quality
All checks passed!
624 files already formatted

pytest tests/trainer/test_step_sync.py tests/distributed/test_torch_compile.py
47 passed, 1 skipped
```

Controlled A-B-B-A validation was performed on the pre-rebase feature commit `469f5c61` based on `main@84bc3601`. The only intervening upstream change before the current head was the unrelated SeedOss documentation PR #1023.

Configuration: 2 x NVIDIA A800 80GB, Qwen3-VL toy model, FSDP2, `dp_replicate_size=2`, `dp_shard_size=1`, gradient accumulation 4, BF16 parameters, FP32 reductions, and dynamic batching enabled.

- Both ranks disabled replica all-reduce at micro-step 0 and restored it at micro-step 3 of every optimizer step.
- Same-rank loss and gradient norm matched exactly across A1/B1/B2/A2.
- Mean median step time decreased from 20.4740 s to 19.8462 s (3.07%); derived throughput increased by 3.16%.
- Peak allocated memory was unchanged; peak reserved memory decreased by 0.287 GiB (1.83%).

The hardware result validates the VLM trainer path and tested two-GPU topology. It does not claim four-GPU scaling, `dp_shard_size > 1`, NPU behavior, or long-horizon convergence.

### API and Usage Example

No API or configuration changes.

### Design & Code Changes

- Call the existing HSDP all-reduce helper after `model_reshard()` in the text, VLM, and DPO micro-batch loops.
- Apply the same lifecycle in DiT training while preserving the offline-embedding path.
- Add a boundary test for the existing helper and verify every trainer train loop invokes it.
- Update the VLM torch-compile test fixture for the shared helper call.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added focused tests
- [x] No documentation update is required because there is no public API, configuration, or workflow change
- [x] Existing CI coverage exercises the modified trainer test files

### CODE REVIEW

**Review by Coach257 — APPROVED** [review]

**Comment by WenzheWang** [comment]
@Coach257 Could you please rerun the cancelled NPU unit test? The latest self-hosted job was cancelled before executing any test steps, while the GPU unit, GPU e2e, and NPU e2e checks have all passed. Thanks.
